### PR TITLE
fix(ir): Validate row reduction scratch shape

### LINKED ISSUE #2082: [Bug] pl.tile.row_max / row_sum silently compute zeros when tmp_tile is smaller than the input
## Background

In pypto-lib, `models/qwen3/14b/bench_softmax_compare.py` on a2a3 has a **precision** problem: a native-DSL reimplementation of the CCE `EpilogueOnlineSoftmax` leaf produced all-zero row maxima, so every probability came out as `exp(0)=1` and the running softmax denominator landed at exactly 7*512=3584 instead of ~150-230.

Root-caused to the tile-level reduction API rather than the kernel: `pl.tile.row_max(tile, tmp_tile)` needs `tmp_tile` to have the SAME shape as the input (`[rows, cols]`). Passing an output-shaped `[rows, 1]` tmp compiles and runs cleanly but silently yields zeros.

Reproduce with:

    python models/qwen3/14b/bench_softmax_compare.py --impl pypto -p a2a3 -d 0

Reproduction environment:

| Component | Version |
|---|---|
| pypto-lib | `fae6826` (branch: `main`) |
| pypto | `8ebddcb` (branch: `main`) — also reproduces on `9598f66` (`codex/pto-buffer-ir`) |
| simpler | `a756969` (detached) |
| ptoas | `0.48` |
| pto-isa | `83d0131` |
| CANN | not detected (npu-smi 26.0.rc1) |

All components are on their pins (simpler == pypto `HEAD:runtime`, pto-isa == `runtime/pto_isa.pin`, ptoas == `PTOAS_VERSION`).

Diagnosis: **pypto** — the op is constructed and lowered without any check that `tmp_tile` is large enough, and the result is silently wrong.

### Component

Python Bindings

### Description

The tile-level reductions `pl.tile.row_max` / `pl.tile.row_sum` take an explicit `tmp_tile` scratch operand. Its required shape is undocumented and unvalidated. If it is smaller than the input tile, compilation succeeds, execution succeeds, and the result is **all zeros** — no error, no warning.

The tensor-level path is unaffected: `pl.row_max(tensor)` internally allocates a full-width scratch (confirmed in the generated `.pto`) and is correct.

Minimal A/B, identical apart from the tmp tile shape (input is `[8, 512]` fp32):

| tmp_tile shape | Result |
|---|---|
| `[8, 1]`   | **FAIL — all zeros** |
| `[8, 64]`  | **FAIL** |
| `[8, 256]` | **FAIL** |
| `[8, 512]` (== input) | PASS |

Generated IR confirms the contract difference — tensor-level emits `pto.trowmax ins(%s, %tmp : ...rows=8, cols=512..., ...rows=8, cols=512...)` whereas the tile-level call faithfully passes the user's `rows=8, cols=1` tmp straight through to `pto.trowmax` with no verifier complaint.

A follow-on symptom that makes this especially confusing to debug: feeding the (zeroed) result into `pl.row_expand_sub(s, m)` yields `s - m` that is *exactly* 0 everywhere, i.e. it looks like the op computed `x - x`.

### Steps to Reproduce

1. Save as `probe.py`:

```python
import pypto.language as pl, torch, math
from golden import TensorSpec, run_jit

TASKS, RP, SL = 128, 16, 512
ROWS = TASKS * RP
SCALE = 1.0 / math.sqrt(128)

@pl.jit
def probe(scores: pl.Tensor[[ROWS, SL], pl.FP32],
          out_m: pl.Out[pl.Tensor[[ROWS, 1], pl.FP32]]) -> pl.Tensor[[ROWS, 1], pl.FP32]:
    for task in pl.spmd(TASKS, name_hint="probe"):
        r0 = task * RP
        # BUG: [8, 1] silently yields zeros; [8, SL] is correct.
        tmp = pl.create_tile([8, 1], dtype=pl.FP32, target_memory=pl.MemorySpace.Vec)
        s = pl.tile.muls(pl.load(scores, [r0, 0], [8, SL]), SCALE)
        m = pl.row_max(s, tmp)
        out_m = pl.store(m, [r0, 0], out_m)
    return out_m

def golden(v):
    s = v["scores"].reshape(TASKS, RP, SL).float() * SCALE
    v["out_m"].reshape(TASKS, RP, 1)[:, :8] = s[:, :8].max(dim=-1, keepdim=True).values

def init():
    s = torch.zeros(TASKS, RP, SL); s[:, :5, :] = torch.randn(TASKS, 5, SL) * 11.3
    return s.reshape(ROWS, SL)

torch.manual_seed(0)
r = run_jit(fn=probe,
            specs=[TensorSpec("scores", [ROWS, SL], torch.float32, init_value=init),
                   TensorSpec("out_m", [ROWS, 1], torch.float32, is_output=True)],
            golden_fn=golden,
            runtime_cfg={"platform": "a2a3", "device_id": 0, "enable_dep_gen": False},
            rtol=5e-3, atol=2e-2)
print("passed=", r.passed)
```

2. Run it from a pypto-lib checkout (so `golden` is importable): `python probe.py`
3. Observe `out_m` FAIL, every element `0.0`.
4. Change the tmp tile to `pl.create_tile([8, SL], ...)` and re-run — passes.

### Expected Behavior

Either:

- **(a)** `pl.tile.row_max` / `row_sum` reject an undersized `tmp_tile` at op construction (or in the IR verifier) with a clear message stating the required shape; **or**
- **(b)** if `[rows, 1]` is in fact the intended contract, codegen honors it and produces correct results.

Either way the required `tmp_tile` shape should be documented — the docstring currently says only "Temporary tile", and the unified wrapper's error message ("row_max on Tile requires tmp_tile argument") gives no shape guidance.

### Actual Behavior

Compiles and runs with no diagnostic; every reduced value is `0.0`.

Note this is easy to hit: the in-tree kernel `models/deepseek/v4/hc_pre.py:296-301` (pypto-lib) uses exactly the `pl.create_tile([COMB_T_TILE, 1], ...)` form as its `row_max` / `row_sum` scratch, with `COMB_T_TILE = 8` — i.e. the same output-shaped tmp that fails here. That usage may be silently affected and is worth auditing.

### Git Commit ID

`8ebddcb87a95c091c02f7f53b0a877c1e3f3d444`

### NPU Kind

Ascend 910B

### Host Platform

Linux (aarch64)

### Additional Context

Related: #1401 (`tile.row_sum` valid_shape propagation).

Found while reimplementing the 5 handwritten CCE attention leaf kernels in native pypto DSL to benchmark DSL-vs-handwritten (`models/qwen3/14b/bench_softmax_compare.py`). Workaround in that bench was to switch to the tensor-level slicing form.

### PR DESCRIPTION
**Pull request by lyfne123**
## Summary
- reject tile row-reduction scratch buffers whose rank or dimensions are provably smaller than the input
- keep padded scratch shapes valid for tensor-to-tile lowering
- update API documentation, examples, and regression coverage for the scratch-shape contract

## Testing
- cmake --build build --parallel
- pytest tests/ut/ir/operators/test_tile_ops.py::TestTileReductionOps -n auto --maxprocesses 8 -q (26 passed)
- full unit suite: 7391 passed, 32 skipped; 2 environment-dependent runtime tests deselected because the installed _task_interface lacks TENSOR_CHILD_MEMORY_OFFSET
- ruff check .
- ruff format --check .

## Related Issues
Fixes #2082

### CODE REVIEW

File: src/ir/op/tile_ops/reduction.cpp [hunk 1]
```diff
@@ -65,6 +66,33 @@ TypePtr DeduceTileRowReductionType(const std::vector<ExprPtr>& args,
...
                          << input_ndim << " dimensions";
 
+  // The PTO row-reduction instructions use tmp_tile as full-size scratch.
+  // An undersized scratch tile compiles but produces silently incorrect
+  // results, so reject it while constructing the op. Larger extents remain
+  // valid because tensor-to-tile lowering intentionally pads short rows.
+  auto tmp_type = As<TileType>(args[1]->GetType());
+  CHECK_SPAN(tmp_type, args[1]->span_)
+      << "The operator " << op_name << " requires tmp_tile to be a TileType, but got "
+      << args[1]->GetType()->TypeName();
```
[the remark this answers is not in the document]
Comment (lyfne123) [reply src/ir/op/tile_ops/reduction.cpp:77]: Fixed in `aa548fcf` - row-reduction scratch dtype is now validated against the input dtype, with regression coverage and the existing FP16 caller corrected.

File: src/ir/op/tile_ops/reduction.cpp [hunk 2]
```diff
@@ -65,6 +66,33 @@ TypePtr DeduceTileRowReductionType(const std::vector<ExprPtr>& args,
...
+      << "tmp_tile shape " << FormatShape(tmp_type->shape_) << " and input shape "
+      << FormatShape(input_shape);
+  for (size_t i = 0; i < input_shape.size(); ++i) {
+    const auto input_extent = GetConstantDimension(input_shape[i]);
+    const auto tmp_extent = GetConstantDimension(tmp_type->shape_[i]);
+    const bool provably_undersized =
+        input_extent && tmp_extent
+            ? *tmp_extent < *input_extent
+            : ProveValidExtentLessEqual(input_shape[i], tmp_type->shape_[i]) == ProofResult::kFalse;
+    CHECK_SPAN(!provably_undersized, args[1]->span_)
```
[the remark this answers is not in the document]
Comment (lyfne123) [reply src/ir/op/tile_ops/reduction.cpp:102]: Fixed in `aa548fcf` - `row_argmax` and `row_argmin` now require an exact scratch shape, and tests reject both undersized and oversized scratch tiles.

**Comment by lyfne123** [comment]
Companion model fix opened as hw-native-sys/pypto-lib#813. It updates both `hc_pre` implementations so row-reduction scratch tiles match the padded input shape required by this PR.

**Review by YunjiQin — APPROVED** [review]
